### DISCUSSION TITLE
New nullability attribute: `DisallowNullMember`

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/csharplang#5778] [Language Ideas]
We already have precondition and postcondition attributes for arguments, but only postcondition ones from members.
`DisallowNullMemberAttribute` notifies compiler that method expects otherwise nullable member to contain non-null value when it's called (member should be assigned or checked in the calling method. same deal as with the arguments).

Example uses:
1) `public` method validates all members and calls a `private` method, that assumes that validation has already been done:
```cs
class C1
{
    string? s;

    public void Init(string s) => this.s = s;

    public void A()
    {
        _ = s ?? throw new InvalidOperationException("Call Init first");
        B();
    }

    [DisallowNullMember(nameof(s))]
    private void B()
    {
        _ = s.Length; // No warning
    }
}
```
Case in the wild: https://github.com/jamesmontemagno/InAppBillingPlugin/blob/a2b3d5651bb69d86a4a46d368715b680b1eac988/src/Plugin.InAppBilling/InAppBilling.android.cs#L255
`UpgradePurchasedSubscriptionAsync` validates that `BillingClient` is not null and calls `UpgradePurchasedSubscriptionInternalAsync`, that uses it without checks.

2. Builder method calls several `private` methods. Each one initializes one member. Some methods use previously initialized members:
```cs
class C2
{
    string? s1;
    string? s2;

    public void Init(string s)
    {
        A();
        B();
    }

    [MemberNotNull(nameof(s1))]
    public void A()
    {
        s1 = string.Empty;
    }

    [DisallowNullMember(nameof(s1))]
    private void B()
    {
        s2 = s1.ToLower(); // No warning
    }
}
```
Case in the wild: https://github.com/dotnet/runtime/blob/a31ed471c8f528fe7ae0237ba38d74ea9af7d0c5/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L143-L147
`BuildHostConfiguration` initializes only `_hostConfiguration`, after which `CreateHostingEnvironment` uses `_hostConfiguration` to initialize `_hostingEnvironment`, and so on.

**Comment by HaloFour** [comment]
The existing NRT attributes are used to inform consumers how the members impact the nullability of inputs/outputs to those members.  This proposal seems to be about an attribute that would only apply to within the class, which seems like a pretty extreme solution.

**Reply by maxkoshevoi** [reply]
Those are the cases I've found today while annotating stuff. Agree, that public method that requires caller to validate member of it's own instance would be pretty strange, so this one is mostly for private or protected methods.

Alternative solution is for compiler to persist information about member nullability to `private` method, if it's called only once. But that seems like a clumsy solution (still should work though).

If there will be no solution to this and people will be forced to use `!`, than if something changes, NRE will follow (as with any other use of `!`). I understand that this is not a very common issue, but it's still a clear blind spot for a compiler that can be addressed.

**Reply by RikkiGibson** [reply]
`[DisallowNullMember]` might fit our name convention more closely.

**Reply by mattwar** [reply]
Maybe `[ItsNotNullNow]`

**Reply by maxkoshevoi** [reply]
`[SwearToGodItsNotNull]` 😄

But if serious, there's already a `[MemberNotNull]` attribute that refers to type member, so having a `Member` word in the name would be good for consistency (and clarity)

**Comment by TahirAhmadov** [comment]
Does this mean in the OP examples, calling `B` generates a warning under certain conditions? If so, what are those conditions? If not, how is this different than using null-suppress `!`?

**Reply by maxkoshevoi** [reply]
> Does this mean in the OP examples, calling `B` generates a warning under certain conditions?

Calling `B` won't generate a warning, but there'll be a warning inside `B` on marked lines that `s` and `s1` respectively could be `null`, since compiler doesn't know that they were already checked by the calling method (`A`).

> how is this different than using null-suppress `!`?

`!` just suppresses the warning. But what if something changes this calling method, and suppression will no longer be valid? You'll get an NRE at runtime. With `[DisallowNullMember]` compiler will notify you that something is wrong at compile time.

**Reply by TahirAhmadov** [reply]
> Calling `B` won't generate a warning, but there'll be a warning inside `B` on marked lines that `s` and `s1` respectively could be `null`, since compiler doesn't know that they were already checked by the calling method (`A`).

You mean there will be a warning today without this attribute? I understand that. My question was, is your idea that calling `B` without checking that `s` is not null would be a warning?

**Reply by maxkoshevoi** [reply]
Sorry, misread the question. Yes, calling `B` from the context where `s may be null` will generate a warning.

**Reply by TahirAhmadov** [reply]
OK it makes perfect sense. Thanks for creating this suggestion.

**Comment by TahirAhmadov** [comment]
Any traction on this? I just ran across this need in my code.
